torch/distributed/_tensor/ops/tensor_ops.py

```diff
@@ -472,3 +484,128 @@ def place(vp: Placement, ip: Placement) -> Placement:
             ],
         )
         return result
+
+
+@register_prop_rule("aten.cat.default")
+def cat_rule(op_schema: OpSchema) -> OutputSharding:
+    # the first arg is a list of input tensors' specs
+    tensor_list_specs = cast(List[DTensorSpec], op_schema.args_schema[0])
+    # ndim will also be the result's ndim
+    ndim = 1
+    for spec in tensor_list_specs:
+        ndim = max(ndim, spec.ndim)
+
+    dim = 0  # default dim = 0
+    if (len(op_schema.args_schema) > 1):
+        dim = op_schema.args_schema[1]
+    # normalize arguments
+    if dim < 0:
+        dim += ndim
+
+    # Unshard all input tensors on cat dim before running einop rule
+    # to avoid _Partial in result.
+    need_reshard = False
+    tensor_list_specs_after = []
+    for spec in tensor_list_specs:
+        if is_tensor_dim_sharded(spec.placements, dim=dim):
+            need_reshard = True
+            tensor_list_specs_after.append(
+                DTensorSpec(
+                    mesh=spec.mesh,
+                    placements=unshard_tensor_dim(spec.placements, dim=dim),
+                    shape=spec.shape,
+                    ndim=spec.ndim,
+                )
+            )
+        else:
+            tensor_list_specs_after.append(spec)
+    tensor_list_specs = tensor_list_specs_after
+
+    # TODO: currently einop rule requires every character
+    # in result notation must have appeared in inputs
+    # so we temporarily design cat notation as
+    # "aij,bij->aij". Once we modify this requirement,
+    # we can switch to the more logically reasonable notation
+    # "aij,bij->cij"
+    alphabet = "abcdefghijklmnopqrstuvwxyz"
+    einop_notation_list = []
+
+    l = len(tensor_list_specs)
+    free_dim = alphabet[l:l + ndim - 1]
+    for i, spec in enumerate(tensor_list_specs):
+        if spec.ndim == ndim:
+            # rewrite concat dim
+            dim_word = free_dim[:dim] + alphabet[i] + free_dim[dim:]
+            einop_notation_list.append(dim_word)
+        else:
+            einop_notation_list.append(alphabet[i])
```

Comment:
is this the empty tensor annotation where it have a single char?

Reply:
Not entirely for empty tensor but empty tensor whose `ndim` is smaller than other tensors. This is for case like concatenating `Tensor([], shape=torch.Size([0]))` with `Tensor([[1, 2], [3, 4]], shape=torch.Size([2, 2]))`.